File: scenario/node/src/checkpointeventlistener.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CloseableAsyncIterable } from '@hyperledger/fabric-gateway';
import { BaseEventListener } from './baseeventlistener';
import { EventListener } from './eventlistener';

export class CheckpointEventListener<T> implements EventListener<T> {
    readonly #checkpoint: (event: T) => Promise<void>;
    #eventListener: BaseEventListener<T>;

    constructor(events: CloseableAsyncIterable<T>, checkpoint: (event: T) => Promise<void>) {
        this.#eventListener = new BaseEventListener<T>(events);
        this.#checkpoint = checkpoint;
    }

    async next(): Promise<T> {
        const event = await this.#eventListener.next();
        await this.#checkpoint(event);
        return event;
    }

    close(): void {
        this.#eventListener.close();
    }
}

File: java/src/test/java/scenario/CheckpointEventListener.java
package scenario;

import java.io.IOException;

import org.hyperledger.fabric.client.CloseableIterator;

public final class CheckpointEventListener<T> implements EventListener<T> {
    private final EventListener<T> eventListener;
    private final CheckpointCall<T> checkpoint;

    @FunctionalInterface
    public interface CheckpointCall<T> {
        void accept(T event) throws IOException;
    }

    CheckpointEventListener(final CloseableIterator<T> iterator, final CheckpointCall<T> checkpoint) {
        eventListener = new BasicEventListener<>(iterator);
        this.checkpoint = checkpoint;
    }

    @Override
    public T next() throws InterruptedException, IOException {
        T event = eventListener.next();
        checkpoint.accept(event);
        return event;
    }

    @Override
    public void close() {
        eventListener.close();
    }
}
